File: crm/fxiaoke_crm_data.js
const axios = require('axios');

/**
 * 纷享销客CRM业务对象数据查询类
 * 提供CRM业务对象的数据查询功能
 */
class FxiaokeCRMDataClient {
    constructor(authManager) {
        this.authManager = authManager;
        
        // 请求配置
        this.requestConfig = {
            baseURL: 'https://open.fxiaoke.com',
            timeout: 15000,
            headers: {
                'Content-Type': 'application/json'
            }
        };
        
        // 重试配置
        this.retryConfig = {
            maxRetries: 3,
            retryDelay: 1000,
            backoffMultiplier: 2
        };
        
        // 频率限制配置（根据API文档：100次/20秒）
        this.rateLimitConfig = {
            maxRequestsPerWindow: 100,
            windowSize: 20000, // 20秒
            requestQueue: [],
            lastRequestTime: 0
        };
    }

    /**
     * 查询CRM业务对象数据
     * @param {Object} params 查询参数
     * @param {string} params.currentOpenUserId 当前用户ID（通过手机号查询员工接口返回）
     * @param {string} params.dataObjectApiName 对象ApiName
     * @param {Array<string>} params.fieldProjection 返回数据中包含的字段，为空则返回所有字段
     * @param {Object} [params.searchQueryInfo] 查询条件
     * @param {number} [params.searchQueryInfo.limit] 查询条数
     * @param {number} [params.searchQueryInfo.offset] 查询开始偏移量
     * @param {Object} [params.searchQueryInfo.filters] 查询条件
     * @param {string} [params.searchQueryInfo.filters.field_name] 字段apiName
     * @param {Array<string>} [params.searchQueryInfo.filters.field_values] 字段值
     * @param {string} [params.searchQueryInfo.filters.operator] 操作符
     * @param {Array<Object>} [params.searchQueryInfo.orders] 排序
     * @param {string} [params.searchQueryInfo.orders.fieldName] 排序字段
     * @param {boolean} [params.searchQueryInfo.orders.isAsc] 是否升序
     * @param {boolean} [params.igonreMediaIdConvert=false] 是否忽略转换MediaId
     * @returns {Promise<Array>} 返回数据数组
     */
    async findSimpleData(params = {}) {
        try {
            // 验证必需参数
            this.validateParams(params);
            
            // 获取认证信息
            const authInfo = await this.authManager.getCorpAccessToken();
            
            // 构建请求数据
            const requestData = {
                corpAccessToken: authInfo.corpAccessToken,
                corpId: authInfo.corpId,
                currentOpenUserId: params.currentOpenUserId,
                data: {
                    dataObjectApiName: params.dataObjectApiName,
                    field_projection: params.fieldProjection || [],
                    igonreMediaIdConvert: params.igonreMediaIdConvert || false,
                    search_query_info: this.buildSearchQueryInfo(params.searchQueryInfo)
                }
            };

            console.log('[FxiaokeCRMData] 请求业务对象数据，参数:', {
                corpId: authInfo.corpId,
                currentOpenUserId: params.currentOpenUserId,
                dataObjectApiName: params.dataObjectApiName,
                fieldProjection: params.fieldProjection?.length || 0,
                limit: params.searchQueryInfo?.limit,
                offset: params.searchQueryInfo?.offset
            });

            // 发送请求
            const response = await this.requestWithRateLimit(
                '/cgi/crm/custom/v2/data/findSimple',
                requestData
            );
            
            console.log('[FxiaokeCRMData] 业务对象数据查询成功');
            
            // 提取数据数组并返回
            if (response.data && response.data.dataList) {
                return response.data.dataList;
            } else {
                console.warn('[FxiaokeCRMData] 响应中没有找到dataList，返回空数组');
                return [];
            }
            
        } catch (error) {
            console.error('[FxiaokeCRMData] 查询业务对象数据失败:', error.message);
            throw error;
        }
    }

    /**
     * 构建查询条件
     * @param {Object} searchQueryInfo 查询条件
     * @returns {Object} 构建后的查询条件
     */
    buildSearchQueryInfo(searchQueryInfo = {}) {
        const result = {};
        
        // 添加基本查询参数
        if (searchQueryInfo.limit !== undefined) {
            result.limit = searchQueryInfo.limit.toString();
        }
        
        if (searchQueryInfo.offset !== undefined) {
            result.offset = searchQueryInfo.offset.toString();
        }
        
        // 添加过滤条件
        if (searchQueryInfo.filters) {
            result.filters = {
                field_name: searchQueryInfo.filters.field_name,
                field_values: Array.isArray(searchQueryInfo.filters.field_values) 
                    ? searchQueryInfo.filters.field_values 
                    : [searchQueryInfo.filters.field_values],
                operator: searchQueryInfo.filters.operator
            };
        }
        
        // 添加排序条件
        if (searchQueryInfo.orders && Array.isArray(searchQueryInfo.orders)) {
            result.orders = searchQueryInfo.orders.map(order => ({
                fieldName: order.fieldName,
                isAsc: order.isAsc ? "true" : "false"
            }));
        }
        
        return result;
    }

    /**
     * 验证必需参数
     * @param {Object} params 参数对象
     */
    validateParams(params) {
        if (!params.currentOpenUserId) {
            throw new Error('currentOpenUserId是必需参数');
        }
        
        if (!params.dataObjectApiName) {
            throw new Error('dataObjectApiName是必需参数');
        }
        
        if (!Array.isArray(params.fieldProjection)) {
            throw new Error('fieldProjection必须是数组类型');
        }
    }

    /**
     * 带频率限制的请求方法
     * @param {string} endpoint API端点
     * @param {Object} data 请求数据
     * @returns {Promise<Object>} API响应
     */
    async requestWithRateLimit(endpoint, data) {
        // 检查频率限制
        await this.checkRateLimit();
        
        // 发送请求
        const response = await this.requestWithRetry(endpoint, data);
        
        // 更新请求时间
        this.rateLimitConfig.lastRequestTime = Date.now();
        
        return response;
    }

    /**
     * 检查频率限制
     */
    async checkRateLimit() {
        const now = Date.now();
        const timeSinceLastRequest = now - this.rateLimitConfig.lastRequestTime;
        
        // 如果距离上次请求时间小于窗口大小，需要等待
        if (timeSinceLastRequest < this.rateLimitConfig.windowSize) {
            const waitTime = this.rateLimitConfig.windowSize - timeSinceLastRequest;
            console.log(`[FxiaokeCRMData] 频率限制，等待 ${waitTime}ms`);
            await this.sleep(waitTime);
        }
    }

    /**
     * 带重试机制的请求
     * @param {string} endpoint API端点
     * @param {Object} data 请求数据
     * @returns {Promise<Object>} API响应
     */
    async requestWithRetry(endpoint, data) {
        let lastError;
        
        for (let attempt = 1; attempt <= this.retryConfig.maxRetries; attempt++) {
            try {
                const response = await axios.post(
                    endpoint,
                    data,
                    this.requestConfig
                );
                
                // 检查API错误
                if (response.data.errorCode !== 0) {
                    throw new Error(`API错误: ${response.data.errorCode} - ${response.data.errorMessage || response.data.errorDescription}`);
                }
                
                return response.data;
                
            } catch (error) {
                lastError = error;
                console.warn(`[FxiaokeCRMData] 第${attempt}次请求失败:`, error.message);
                
                // 如果是最后一次尝试，抛出错误
                if (attempt === this.retryConfig.maxRetries) {
                    break;
                }
                
                // 计算延迟时间（指数退避）
                const delay = this.retryConfig.retryDelay * Math.pow(this.retryConfig.backoffMultiplier, attempt - 1);
                console.log(`[FxiaokeCRMData] ${delay}ms后重试...`);
                await this.sleep(delay);
            }
        }
        
        throw new Error(`请求失败，已重试${this.retryConfig.maxRetries}次: ${lastError.message}`);
    }

    /**
     * 查询客户数据
     * @param {Object} params 查询参数
     * @param {string} params.currentOpenUserId 当前用户ID
     * @param {Array<string>} [params.fields] 返回字段列表
     * @param {number} [params.limit=10] 查询条数
     * @param {number} [params.offset=0] 查询偏移量
     * @param {Object} [params.filters] 过滤条件
     * @param {Array<Object>} [params.orders] 排序条件
     * @returns {Promise<Array>} 客户数据数组
     */
    async findCustomerData(params = {}) {
        return this.findSimpleData({
            ...params,
            dataObjectApiName: 'Customer',
            fieldProjection: params.fields || ['_id', 'name', 'mobile', 'email', 'last_modified_time']
        });
    }

    /**
     * 查询联系人数据
     * @param {Object} params 查询参数
     * @param {string} params.currentOpenUserId 当前用户ID
     * @param {Array<string>} [params.fields] 返回字段列表
     * @param {number} [params.limit=10] 查询条数
     * @param {number} [params.offset=0] 查询偏移量
     * @param {Object} [params.filters] 过滤条件
     * @param {Array<Object>} [params.orders] 排序条件
     * @returns {Promise<Array>} 联系人数据数组
     */
    async findContactData(params = {}) {
        return this.findSimpleData({
            ...params,
            dataObjectApiName: 'Contact',
            fieldProjection: params.fields || ['_id', 'name', 'mobile', 'email', 'last_modified_time']
        });
    }

    /**
     * 查询销售机会数据
     * @param {Object} params 查询参数
     * @param {string} params.currentOpenUserId 当前用户ID
     * @param {Array<string>} [params.fields] 返回字段列表
     * @param {number} [params.limit=10] 查询条数
     * @param {number} [params.offset=0] 查询偏移量
     * @param {Object} [params.filters] 过滤条件
     * @param {Array<Object>} [params.orders] 排序条件
     * @returns {Promise<Array>} 销售机会数据数组
     */
    async findOpportunityData(params = {}) {
        return this.findSimpleData({
            ...params,
            dataObjectApiName: 'Opportunity',
            fieldProjection: params.fields || ['_id', 'name', 'amount', 'stage', 'last_modified_time']
        });
    }

    /**
     * 根据时间范围查询数据
     * @param {Object} params 查询参数
     * @param {string} params.currentOpenUserId 当前用户ID
     * @param {string} params.dataObjectApiName 对象ApiName
     * @param {Array<string>} params.fieldProjection 返回字段列表
     * @param {Date|string|number} params.startTime 开始时间
     * @param {Date|string|number} params.endTime 结束时间
     * @param {string} [params.timeField='last_modified_time'] 时间字段名
     * @param {number} [params.limit=10] 查询条数
     * @param {number} [params.offset=0] 查询偏移量
     * @returns {Promise<Array>} 查询结果数组
     */
    async findDataByTimeRange(params = {}) {
        const startTime = this.convertToTimestamp(params.startTime);
        const endTime = this.convertToTimestamp(params.endTime);
        
        return this.findSimpleData({
            currentOpenUserId: params.currentOpenUserId,
            dataObjectApiName: params.dataObjectApiName,
            fieldProjection: params.fieldProjection,
            searchQueryInfo: {
                limit: params.limit || 10,
                offset: params.offset || 0,
                filters: {
                    field_name: params.timeField || 'last_modified_time',
                    field_values: [startTime.toString(), endTime.toString()],
                    operator: 'BETWEEN'
                },
                orders: [
                    {
                        fieldName: params.timeField || 'last_modified_time',
                        isAsc: false
                    }
                ]
            }
        });
    }

    /**
     * 转换时间为时间戳
     * @param {Date|string|number} time 时间
     * @returns {number} 时间戳
     */
    convertToTimestamp(time) {
        if (typeof time === 'number') {
            return time;
        }
        
        if (typeof time === 'string') {
            return new Date(time).getTime();
        }
        
        if (time instanceof Date) {
            return time.getTime();
        }
        
        throw new Error('无效的时间格式');
    }

    /**
     * 设置重试配置
     * @param {Object} config 重试配置
     */
    setRetryConfig(config) {
        this.retryConfig = { ...this.retryConfig, ...config };
    }

    /**
     * 设置频率限制配置
     * @param {Object} config 频率限制配置
     */
    setRateLimitConfig(config) {
        this.rateLimitConfig = { ...this.rateLimitConfig, ...config };
    }

    /**
     * 设置请求超时时间
     * @param {number} timeout 超时时间（毫秒）
     */
    setTimeout(timeout) {
        this.requestConfig.timeout = timeout;
    }

    /**
     * 休眠函数
     * @param {number} ms 休眠时间（毫秒）
     */
    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    /**
     * 销毁实例
     */
    destroy() {
        // 清理资源
        this.rateLimitConfig.requestQueue = [];
    }
}

module.exports = FxiaokeCRMDataClient; 

File: test_agent/fxiaoke_crm_data.py
import requests
import time
import logging
from typing import Dict, List, Optional, Any, Union
from datetime import datetime, timedelta
from .fxiaoke_auth import FxiaokeAuthManager

logger = logging.getLogger(__name__)


class FxiaokeCRMDataClient:
    """
    纷享销客CRM业务对象数据查询类
    提供CRM业务对象的数据查询功能
    """
    
    def __init__(self, auth_manager: FxiaokeAuthManager):
        """
        初始化CRM数据客户端
        
        Args:
            auth_manager: 认证管理器实例
        """
        self.auth_manager = auth_manager
        
        # 请求配置
        self.request_config = {
            'baseURL': 'https://open.fxiaoke.com',
            'timeout': 15,
            'headers': {
                'Content-Type': 'application/json'
            }
        }
        
        # 重试配置
        self.retry_config = {
            'maxRetries': 3,
            'retryDelay': 1000,
            'backoffMultiplier': 2
        }
        
        # 频率限制配置（根据API文档：100次/20秒）
        self.rate_limit_config = {
            'maxRequestsPerWindow': 100,
            'windowSize': 20,  # 20秒
            'requestQueue': [],
            'lastRequestTime': 0
        }
    
    async def find_simple_data(
        self,
        currentOpenUserId: str,
        dataObjectApiName: str,
        fieldProjection: Optional[List[str]] = None,
        searchQueryInfo: Optional[Dict[str, Any]] = None,
        ignoreMediaIdConvert: bool = False,
        **kwargs
    ) -> List[Dict[str, Any]]:
        """
        通用查询CRM任意对象数据
        Args:
            currentOpenUserId: 当前用户ID
            dataObjectApiName: 对象ApiName
            fieldProjection: 返回字段列表
            searchQueryInfo: 查询条件，需要通过build_search_query_info方法构建
            ignoreMediaIdConvert: 是否忽略MediaId转换
            kwargs: 其他可选参数
        Returns:
            数据列表
        """
        if not currentOpenUserId:
            raise ValueError('currentOpenUserId是必需参数')
        if not dataObjectApiName:
            raise ValueError('dataObjectApiName是必需参数')

        auth_info = await self.auth_manager.get_corp_access_token()
        data = {
            'dataObjectApiName': dataObjectApiName,
            'ignoreMediaIdConvert': ignoreMediaIdConvert
        }
        if fieldProjection is not None:
            data['field_projection'] = fieldProjection
        if searchQueryInfo is not None:
            data['search_query_info'] = searchQueryInfo

        request_data = {
            'corpAccessToken': auth_info['corpAccessToken'],
            'corpId': auth_info['corpId'],
            'currentOpenUserId': currentOpenUserId,
            'data': data
        }

        logger.info('[FxiaokeCRMData] 请求业务对象数据，参数: %s', request_data)
        response = await self.request_with_rate_limit(
            '/cgi/crm/custom/v2/data/findSimple',
            request_data
        )
        if response.get('code', 0) != 0:
            logger.error(f"[FxiaokeCRMData] API返回错误: {response.get('msg')}")
            raise Exception(f"API Error: {response.get('msg')}")
        return response.get('data', {}).get('dataList', [])
    
    def build_search_query_info(self, search_query_info: Dict[str, Any] = None) -> Dict[str, Any]:
        """
        构建查询条件
        
        Args:
            search_query_info: 查询条件
            
        Returns:
            构建后的查询条件
        """
        result = {}
        search_query_info = search_query_info or {}
        
        # 添加基本查询参数
        if search_query_info.get('limit') is not None:
            result['limit'] = str(search_query_info['limit'])
        
        if search_query_info.get('offset') is not None:
            result['offset'] = str(search_query_info['offset'])
        
        # 添加过滤条件
        if search_query_info.get('filters'):
            filters = search_query_info['filters']
            result['filters'] = {
                'field_name': filters.get('field_name'),
                'field_values': filters['field_values'] if isinstance(filters['field_values'], list) else [filters['field_values']],
                'operator': filters.get('operator')
            }
        
        # 添加排序条件
        if search_query_info.get('orders') and isinstance(search_query_info['orders'], list):
            result['orders'] = [
                {
                    'fieldName': order['fieldName'],
                    'isAsc': "true" if order.get('isAsc') else "false"
                }
                for order in search_query_info['orders']
            ]
        
        return result
    
    def validate_params(self, params: Dict[str, Any]):
        """
        验证必需参数
        
        Args:
            params: 参数对象
        """
        if not params:
            raise ValueError('参数不能为空')
        
        if not params.get('currentOpenUserId'):
            raise ValueError('currentOpenUserId是必需参数')
        
        if not params.get('dataObjectApiName'):
            raise ValueError('dataObjectApiName是必需参数')
        
        if not isinstance(params.get('fieldProjection', []), list):
            raise ValueError('fieldProjection必须是数组类型')
    
    async def request_with_rate_limit(self, endpoint: str, data: Dict[str, Any]) -> Dict[str, Any]:
        """
        带频率限制的请求方法
        
        Args:
            endpoint: API端点
            data: 请求数据
            
        Returns:
            API响应字典
        """
        # 检查频率限制
        await self.check_rate_limit()
        
        # 发送请求
        response = await self.request_with_retry(endpoint, data)
        
        # 更新请求时间
        self.rate_limit_config['lastRequestTime'] = time.time()
        
        return response
    
    async def check_rate_limit(self):
        """检查频率限制"""
        now = time.time()
        time_since_last_request = now - self.rate_limit_config['lastRequestTime']
        
        # 如果距离上次请求时间小于窗口大小，需要等待
        if time_since_last_request < self.rate_limit_config['windowSize']:
            wait_time = self.rate_limit_config['windowSize'] - time_since_last_request
            logger.info(f'[FxiaokeCRMData] 频率限制，等待 {wait_time}秒')
            await self.sleep(wait_time)
    
    async def request_with_retry(self, endpoint: str, data: Dict[str, Any]) -> Dict[str, Any]:
        """
        带重试机制的请求
        
        Args:
            endpoint: API端点
            data: 请求数据
            
        Returns:
            API响应字典
        """
        last_error = None
        
        for attempt in range(1, self.retry_config['maxRetries'] + 1):
            try:
                url = f"{self.request_config['baseURL']}{endpoint}"
                
                response = requests.post(
                    url,
                    json=data,
                    headers=self.request_config['headers'],
                    timeout=self.request_config['timeout']
                )
                
                response.raise_for_status()
                return response.json()
                
            except Exception as error:
                last_error = error
                logger.warning(f'[FxiaokeCRMData] 第{attempt}次请求失败: {str(error)}')
                
                # 如果是最后一次尝试，抛出错误
                if attempt == self.retry_config['maxRetries']:
                    break
                
                # 计算延迟时间（指数退避）
                delay = self.retry_config['retryDelay'] * (self.retry_config['backoffMultiplier'] ** (attempt - 1))
                logger.info(f'[FxiaokeCRMData] {delay}ms后重试...')
                await self.sleep(delay / 1000)  # 转换为秒
        
        raise Exception(f"请求失败，已重试{self.retry_config['maxRetries']}次: {str(last_error)}")
    
    async def find_data_by_time_range(self, params: Dict[str, Any] = None) -> List[Dict[str, Any]]:
        """
        根据时间范围查询数据
        
        Args:
            params: 查询参数
                - currentOpenUserId: 当前用户ID
                - dataObjectApiName: 对象ApiName
                - fieldProjection: 返回字段列表
                - startTime: 开始时间
                - endTime: 结束时间
                - timeField: 时间字段名（默认'last_modified_time'）
                - limit: 查询条数
                - offset: 查询偏移量
                
        Returns:
            数据列表
        """
        try:
            params = params or {}
            
            # 转换时间格式
            start_time = self.convert_to_timestamp(params.get('startTime'))
            end_time = self.convert_to_timestamp(params.get('endTime'))
            time_field = params.get('timeField', 'last_modified_time')
            
            # 构建时间范围过滤条件
            search_query_info = {
                'limit': params.get('limit'),
                'offset': params.get('offset'),
                'filters': {
                    'field_name': time_field,
                    'field_values': [start_time, end_time],
                    'operator': 'BETWEEN'
                },
                'orders': [
                    {
                        'fieldName': time_field,
                        'isAsc': False
                    }
                ]
            }
            
            # 构建完整查询参数
            query_params = {
                'currentOpenUserId': params['currentOpenUserId'],
                'dataObjectApiName': params['dataObjectApiName'],
                'fieldProjection': params.get('fieldProjection', []),
                'searchQueryInfo': search_query_info
            }
            
            return await self.find_simple_data(query_params['currentOpenUserId'], query_params['dataObjectApiName'], query_params['fieldProjection'], query_params['searchQueryInfo'])
            
        except Exception as error:
            logger.error(f'[FxiaokeCRMData] 根据时间范围查询数据失败: {str(error)}')
            raise
    
    def convert_to_timestamp(self, time_value: Union[datetime, str, int, float]) -> int:
        """
        转换时间为时间戳
        
        Args:
            time_value: 时间值，可以是datetime对象、字符串、时间戳
            
        Returns:
            时间戳（毫秒）
        """
        if isinstance(time_value, datetime):
            return int(time_value.timestamp() * 1000)
        elif isinstance(time_value, str):
            # 尝试解析字符串时间
            try:
                dt = datetime.fromisoformat(time_value.replace('Z', '+00:00'))
                return int(dt.timestamp() * 1000)
            except ValueError:
                raise ValueError(f'无法解析时间字符串: {time_value}')
        elif isinstance(time_value, (int, float)):
            # 如果是秒级时间戳，转换为毫秒
            if time_value < 10000000000:  # 秒级时间戳
                return int(time_value * 1000)
            else:  # 毫秒级时间戳
                return int(time_value)
        else:
            raise ValueError(f'不支持的时间格式: {type(time_value)}')
    
    def set_retry_config(self, config: Dict[str, Any]):
        """
        设置重试配置
        
        Args:
            config: 重试配置字典
        """
        self.retry_config.update(config)
    
    def set_rate_limit_config(self, config: Dict[str, Any]):
        """
        设置频率限制配置
        
        Args:
            config: 频率限制配置字典
        """
        self.rate_limit_config.update(config)
    
    def set_timeout(self, timeout: int):
        """
        设置请求超时时间
        
        Args:
            timeout: 超时时间（秒）
        """
        self.request_config['timeout'] = timeout
    
    @staticmethod
    async def sleep(seconds: float):
        """
        异步睡眠
        
        Args:
            seconds: 睡眠秒数
        """
        import asyncio
        await asyncio.sleep(seconds)
    
    def destroy(self):
        """清理资源"""
        logger.info('[FxiaokeCRMData] 清理CRM数据客户端资源')
        # 清理请求队列
        self.rate_limit_config['requestQueue'].clear() 